nixos/modules/services/backup/syncoid.nix

```diff
@@ -301,42 +227,103 @@ in
 
   config = mkIf cfg.enable {
-    users = {
-      users = mkIf (cfg.user == "syncoid") {
-        syncoid = {
-          group = cfg.group;
-          isSystemUser = true;
-          # For syncoid to be able to create /var/lib/syncoid/.ssh/
-          # and to use custom ssh_config or known_hosts.
-          home = "/var/lib/syncoid";
-          createHome = false;
-        };
-      };
-      groups = mkIf (cfg.group == "syncoid") {
-        syncoid = { };
-      };
-    };
+    assertions = [
+      {
+        assertion = cfg.nftables.enable -> config.networking.nftables.enable;
+        message = "config.networking.nftables.enable must be set when config.services.syncoid.nftables.enable is set";
+      }
+    ];
 
     systemd.services = mapAttrs'
       (name: c:
+        let
+          sshKeyCred = builtins.split ":" c.sshKey;
+        in
         nameValuePair "syncoid-${escapeUnitName name}" (mkMerge [
           {
             description = "Syncoid ZFS synchronization from ${c.source} to ${c.target}";
             after = [ "zfs.target" ];
             startAt = cfg.interval;
-            # syncoid may need zpool to get feature@extensible_dataset
-            path = [ "/run/booted-system/sw/bin/" ];
+            # Here we explicitly use the booted system to guarantee the stable API needed by ZFS.
+            # Moreover syncoid may need zpool to get feature@extensible_dataset.
+            path = [ "/run/booted-system/sw" ];
+            # Prevents missing snapshots during DST changes
+            environment.TZ = "UTC";
+            # A custom LD_LIBRARY_PATH is needed to access in `getent passwd`
+            # the systemd's entry about the DynamicUser=,
+            # so that ssh won't fail with: "No user exists for uid $UID".
+            environment.LD_LIBRARY_PATH = config.system.nssModules.path;
             serviceConfig = {
               ExecStartPre =
-                (map (buildAllowCommand c.localSourceAllow) (localDatasetName c.source)) ++
-                (map (buildAllowCommand c.localTargetAllow) (localDatasetName c.target));
+                # Recursively remove any residual permissions
+                # given on local+descendant datasets (source, target or target's parent)
+                # to any currently unknown (hence unused) systemd dynamic users (UID/GID range 61184…65519),
+                # which happens when a crash has occurred
+                # during any previous run of a syncoid-*.service (not only this one).
+                map
+                  (dataset:
+                    "+" + pkgs.writeShellScript "zfs-unallow-unused-dynamic-users" ''
+                      set -eu
+                      zfs allow "$1" |
+                      sed -ne 's/^\t\(user\|group\) (unknown: \([0-9]\+\)).*/\1 \2/p' |
+                      {
+                        declare -a uids
+                        while read -r role id; do
+                          if [ "$id" -ge 61184 ] && [ "$id" -le 65519 ]; then
```

Comment:
Sounds like a good idea. I guess it would first deserve its own PR to expose `DYNAMIC_UID_MIN`/`MAX`, and then another PR to use them in `services.syncoid`.